lib/fetcher/modulator.ts

```diff
@@ -0,0 +1,307 @@
+            // pass
+        }
+
+        this.ranker = ranker;
+        this.notifier = notifier;
+        this.factory = factory;
+        for (const item of readd) {
+            this.push(item.item);
+        }
+    }
+
+    length(): number {
+        return this.state.item.todo.length;
+    }
+
+    push(item: T) {
+        const indexed = { item, index: this.index };
+        this.state.item.todo.push(indexed);
+        this.index += 1;
+        this.ranker.push(indexed);
+        this.checkReady();
+    }
+
+    seen(url: string): boolean {
+        return this.state.item.immutable.has(url);
+    }
+
+    recordMutable(url: string, item: T): void {
+        this.state.item.mutable.set(url, item);
+    }
+
+    recordImmutable(url: string): void {
+        this.state.item.immutable.add(url);
+        // Remove from mutable list
+        this.state.item.mutable.delete(url);
+    }
+
+    recordEmitted(url: string): void {
+        this.state.item.emitted.add(url);
+        // Remove form unemitted list
+        if (this.state.item.unemitted.has(url)) {
+            this.state.item.unemitted.delete(url);
+        }
+    }
+
+    recordUnemitted(url: string, member: M): void {
+        this.state.item.unemitted.set(url, member);
+    }
+
+    getTodo(): ReadonlyArray<T> {
+        return this.state.item.todo.map((x) => x.item);
+    }
+
+    getInFlight(): ReadonlyArray<T> {
+        return this.state.item.inflight.map((x) => x.item);
+    }
+
+    getMutable(): ReadonlyMap<string, T> {
+        return <ReadonlyMap<string, T>>this.state.item.mutable;
+    }
+
+    getUnemitted(): Array<M> {
+        return Array.from(this.state.item.unemitted.values());
+    }
+
+    getEmitted(): ReadonlySet<string> {
+        return <ReadonlySet<string>>this.state.item.emitted;
+    }
+
+    finished(index: number) {
+        const removeIdx = this.state.item.inflight.findIndex(
+            (x) => x.index == index,
+        );
+        if (removeIdx >= 0) {
+            this.state.item.inflight.splice(removeIdx, 1);
+        } else {
+            this.logger.error(
+                "[finished] Expected to be able to remove inflight item",
+            );
+        }
+
+        this.at -= 1;
+        this.checkReady();
+    }
+
+    checkReady() {
+        if (this.factory.paused) {
+            return;
+        }
+
+        while (this.at < this.factory.concurrent) {
+            const item = this.ranker.pop();
+            if (item) {
+                // This item is no longer todo
+                // I'm quite afraid to use filter for this
+                const removeIdx = this.state.item.todo.findIndex(
+                    (x) => x.index == item.index,
+                );
+                if (removeIdx >= 0) {
+                    this.state.item.todo.splice(removeIdx, 1);
+                } else {
+                    this.logger.error(
+                        "[checkReady] Expected to be able to remove inflight item",
```

Comment:
This is expected tobe able to remove todo item, correct?

Reply:
Indeed. Fixed in https://github.com/rdf-connect/ldes-client/pull/94/commits/053a102f0a8c53870e9c4c710ae49c2dfb63e1fc